lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+# frozen_string_literal: true
+
+module RGeo
+  module Cartesian
+    module Graphs
+      # HalfEdge represents an edge as 2 directed edges.
+      # One half-edge will have it's origin at edge.s, the other
+      # at edge.e. Both half-edges will be linked as each other's twins.
+      #
+      # HalfEdges also contain references to the next and prev half-edges,
+      # where next's origin is this half-edges destination. Prev's destination
+      # is this half-edge's origin.
+      class HalfEdge
+        include Comparable
+
+        # Creates 2 half edges from an edge.
+        # They will be assigned as each other's twins.
+        # The Half Edges will be returned in the order of points
+        # of the edge (start, end).
+        #
+        # @param edge [RGeo::Cartesian::Segment]
+        #
+        # @return [Array]
+        def self.from_edge(edge)
+          # TODO: should we raise an error if the edge is degenerate?
```

Comment:
I guess so :)

Reply:
I guess my concern with raising an error is that the context this will be created in is when we're checking for validity (possibly more later). So we'd have to make sure that it is handled properly upstream. I'll think about it more, but it probably does make sense to raise and we'll just need to keep track of all the errors.

Reply:
After thinking about it more, I think it's fine to just leave it. It shouldn't affect anything else in the graph and just be disconnected from everything entirely. I'll add a test for it though.